src/Interpreters/Cache/QueryConditionCache.cpp

```diff
@@ -125,49 +128,74 @@ QueryConditionCacheWriter::~QueryConditionCacheWriter()
 
 void QueryConditionCacheWriter::addRanges(const UUID & table_id, const String & part_name, const MarkRanges & mark_ranges, size_t marks_count, bool has_final_mark)
 {
-    QueryConditionCache::Key key = {table_id, part_name, condition_hash, condition};
-
-    std::lock_guard<std::mutex> lock(mutex);
+    /// Clickhouse scan ranges in parallel, and the addRanges method will be frequently called.
+    /// Extra care is required to avoid performance problems caused by lock contention.
 
-    bool is_insert;
+    QueryConditionCache::Key key = {table_id, part_name, condition_hash, condition};
 
-    /// ClickHouse scans ranges within the same part in parallel.
-    /// The first scan thread which calls addRanges creates and inserts an entry.
-    /// The other scan threads update existing entry and update it.
-    if (auto it = new_entries.find(key); it == new_entries.end())
+    CacheEntryPtr cache_entry;
     {
-        /// By default, all marks potentially are potential matches, i.e. we can't skip them.
-        /// Treat all marks for the new entry of the part as potential matches, i.e. don't skip them during read.
-        /// This is important for error handling: Imagine an exception is thrown during query execution and the stack is unwound. At that
-        /// point, a new entry may not have received updates for all scanned ranges within the part. As a result, future scans queries could
-        /// skip too many ranges, causing wrong results. This situation is prevented by initializing all marks of each entry as non-matching.
-        /// Even if there is an exception, future scans will not skip them.
-        QueryConditionCache::Entry entry(marks_count, true);
+        std::lock_guard lock(mutex);
```

Comment:
Here, we are unconditionally acquiring the lock in exclusive mode.
Is there any way this can be avoided or deferred?